### ISSUE TITLE
Refactoring the component

### ISSUE THREAD

**Issue by jean-michelet** [fastify/fastify-autoload#383]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the issue has not already been raised

### Issue

The code is becoming increasingly difficult to maintain, I think it deserves a redesign as suggested [here](https://github.com/fastify/fastify-autoload/issues/380#issuecomment-2159851069) and [here](https://github.com/fastify/fastify-autoload/pull/381#discussion_r1635303218).

I'd like to know if you have any suggestions before proceeding, we could also consider proceeding step by step.

**Comment by Uzlopak** [comment]
I dont see the maintenance problem. The code is actually of very high quality. Also it is "only" about 500 LOC. The issue is that the problem it tries to solve is complex itself. That results in not seeing every "edge" case.

**Comment by jean-michelet** [comment]
Perhaps I should have expressed myself less assertively.
I have the feeling that the code could be better organized and more readable, but this not the reason I want to invest time in it. 
Would refactoring help solve the issue problem and eventually add new features? If yes, how?

Maybe you can elaborate what you were thinking by saying this @climba03003:
> Rework the files finding and tree building - it is the critical one because all the information is scattered currently.

**Comment by climba03003** [comment]
The main problem is that the current information split into two array.
https://github.com/fastify/fastify-autoload/blob/13706979a74cfac4c414d9114480c04f2c52833d/index.js#L26-L27
Sometimes the information is not about to find inside some function.

For what I propose is that we should define the works into 3 clear steps.
1. Finds all available files including the config.
2. Build the tree based on step 1. and the plugin configure
3. Registered plugin based on the tree in step 2.

In this case we can always change the function inside step 2. to provide different behavior instead of looking around different function and gather necessary information in different place.

**Comment by jean-michelet** [comment]
I agree that hooks should be composed to plugins before registration step.

> Build the tree based on step 1. and the plugin config

Btw, do you think we should allow `auto` exports from autohooks files too?
```ts
export const autoConfig = { name: 'y' }
export const autoPrefix = "/prefixed";
```

I am not sure that it's a good idea but some people want to use them there, I saw this in a [reproduction issue repo](https://github.com/rommelandrea/fastify-autoload-bug/blob/a2720e87a4c5d6fbfc6029868989774e525d1da8/routes/example/autohooks.js#L7).

If we don't maybe, we should throw a clean error to educate users about it.

**Comment by climba03003** [comment]
> Btw, do you think we should allow auto exports from autohooks files too?

It is not a good idea to permit hooks having different config or prefix than the `index` plugin.
The main reason is that it is very easy to misconfig, for example you may experience some path only contain hooks but no route.

**Comment by jean-michelet** [comment]
Do #401 opened more possibilities to fix #326? I will investigate this week but maybe you have some feedbacks since.
